src/modules/formBuilder/components/FormBuilder.tsx

```diff
@@ -0,0 +1,174 @@
+  lastUpdatedBy?: string;
+  selectedItem?: FormItem;
+  setSelectedItem: Dispatch<SetStateAction<FormItem | undefined>>;
+  originalLinkId?: string;
+  setOriginalLinkId: Dispatch<SetStateAction<string | undefined>>;
+  closeItemEditor: () => void;
+}
+
+const FormBuilder: React.FC<FormBuilderProps> = ({
+  formDefinition,
+  workingDefinition,
+  setWorkingDefinition,
+  errorState,
+  onSave,
+  saveLoading,
+  lastUpdatedDate,
+  lastUpdatedBy,
+  selectedItem,
+  setSelectedItem,
+  originalLinkId,
+  setOriginalLinkId,
+  closeItemEditor,
+}) => {
+  const dirty = useMemo(() => {
+    return !isEqual(workingDefinition, formDefinition.definition);
+  }, [workingDefinition, formDefinition.definition]);
+
+  if (!workingDefinition || !setWorkingDefinition) return <Loading />;
+
+  return (
+    <Box sx={{ display: 'flex' }}>
+      <FormBuilderPalette
+        onItemClick={(itemType) => {
+          const newItem: FormItem = {
+            linkId: crypto.randomUUID(), // TODO(#6083) - this is a placeholder, but right now will only work in localhost
+            text: itemType.toString(),
+            type: itemType,
+            required: false,
+            warnIfEmpty: false,
+            hidden: false,
+            readOnly: false,
+            repeats: false,
+            prefill: false,
+            disabledDisplay: DisabledDisplay.Hidden,
+            enableBehavior: EnableBehavior.Any,
+          };
+
+          const newDefinition: FormDefinitionJson = {
+            ...workingDefinition,
+            item: [...workingDefinition.item, newItem],
+          };
+
+          setWorkingDefinition(newDefinition);
+        }}
+      />
+      {selectedItem && originalLinkId && (
+        <FormItemEditor
+          selectedItem={selectedItem}
+          originalLinkId={originalLinkId}
+          definition={formDefinition}
+          saveLoading={saveLoading}
+          errorState={errorState}
+          onSave={(item, originalLinkId) => {
+            const newDefinition = updateFormItem(
+              workingDefinition,
+              item,
+              originalLinkId
+            );
+
+            onSave(newDefinition);
+          }}
+          onDiscard={closeItemEditor}
+        />
+      )}
+      <Box
+        sx={
+          // Padding matches the padding usually applied in DashboardContentContainer.
+          // (It's moved in here because of the drawer)
+          {
+            flexGrow: 1,
+            pt: 2,
+            pb: 8,
+            px: { xs: 1, sm: 3, lg: 4 },
+          }
+        }
+      >
+        <FormBuilderHeader
+          formDefinition={formDefinition}
+          lastUpdatedDate={lastUpdatedDate}
+        />
+        <Box sx={{ p: 4 }}>
+          <FormTree
+            definition={workingDefinition}
+            onEditClick={(item: FormItem) => {
+              // todo @martha - this should NOT auto-save, it should instead prompt to save
+              // onSave(workingDefinition).then((values) => {
```

Comment:
@gigxz let's discuss the prompt-to-save approach a bit more? Here's what I learned so far:

react-router-dom has some handy built-in functionality for this, `useBlocker`. I played around with it and got something basic running [here](https://github.com/greenriver/hmis-frontend/commit/68385b327771da277b667146bcf25db4ac52420e). But it has some major caveats:
- Most importantly: It doesn’t work if you just plop it inside of a `<BrowserRouter/>`. It needs a Data Router: https://reactrouter.com/en/6.23.1/routers/picking-a-router 
    - This would be a big refactor of our routing, and it definitely feels like scope creep to me.
- It also only works for navigation events within our react app. It does not work for “document events” like hitting reload or navigating to another site directly. https://reactrouter.com/en/main/hooks/use-blocker (It does work on the back button, I think because React Router hijacks session history. It doesn't work on logout, which I wasn't clear about.)
- Finally, they state outright in their docs that it’s an anti-pattern to hijack user navigation, and recommend using session storage instead to persist unsaved values. Given the lift to get this implemented, maybe that would be easier too.
So, I guess, 3 proposed ideas:
- Full steam ahead with upgrading our React Router, which they [recommend doing anyway](https://reactrouter.com/en/6.23.1/routers/picking-a-router#web-projects)
- Accept a non-styled alert component, and attach it directly to the `window.beforeunload` event https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event 
- Just use session storage

Reply:
Thanks Martha. I thought we landed on a lower-lift approach that would just present the save modal when you clicked on certain elements on the page ("preview", "edit item"), and not worry about intercepting browser navigation.

All 3 of the proposed options seem too big for https://github.com/open-path/Green-River/issues/6103. It would be good to capture your research somewhere, though, can you create a ticket for "Implement blocker to stop navigation when user has unsaved changes" so we can address it for Assessments and Form building together? thanks!

Reply:
Ah, sorry, I must have misunderstood our conclusion on the design session, but the lower lift approach sounds fine to me for now. Yes, I will create a ticket.